stdenv: add `stdenv.isAvailable pkg`

amjoseph-nixpkgs opened the pull request:
#### Motivation

The current `meta.availableOn` has two problems:

- It is "shallow" -- a package is "available on" a platform even if its `buildInputs` are not
- It only knows about the `hostPlatform`.  For example, a package is "available" even when:
  - the compiler being used cannot emit code for the `hostPlatform`
  - the compiler being used cannot be built for the `buildPlatform`
  - the package's build process assumes it can execute `hostPlatform` binaries but `!buildPlatform.canExecute hostPlatform`

This PR fixes both problems.

##### Specific examples of multi-platform constraints
- Some packages cannot build unless `buildPlatform.canExecute hostPlatform`, and we have no clean way to express this.  Often (`xdg-open`, `p11-kit`, and others) these packages are *optional* dependencies, and the depending package needs to detect their unavailability so they can be omitted rather than simply failing the build.
- LLVM can't be used if `targetPlatform.isMips64n32`, but GCC can
- `go` is missing bootstrap binaries for a few platforms yet can generate code for those platforms, so we can cross-compile to these platforms but can't build a native compiler for them.
- `ghc` is missing usable bootstrap binaries for `powerpc64*`, although it can emit code for (and even compile itself for) that platform.
#### Description of changes

This commit adds `meta.availability.{build,host,target}.{bad,only}`, which are computed automatically, and updates `meta.availableOn` to use these attributes.

- For the hostPlatform this is done in a backwards-compatible way:    `meta.availability.host` incorporates    `meta.{platforms,badPlatforms}` from the package as well as `meta.availability.host.{bad,only}` from its `buildInputs`.  Because the constraints are cumulative at each package, deep recursion (like checkMetaRecursively) is not needed; each package's `meta.availability` only looks at that package's immediate inputs.

- For the buildPlatform and targetPlatform two additional package-provided `meta` attributes are needed: `meta.badBuildPlatforms` and `meta.badTargetPlatforms`.  The latter is needed only for compiler-like packages.
  - I'm not sure `badBuildPlatforms` is useful; I don't have an example for it yet.  However `meta.availability.build.bad` is definitely useful when we have a compiler written in its own language that can generate code for a platform but upstream doesn't provide "bootstrap binaries" for that platform.  Two real-life examples: GHC on powerpc64le and golang on all flavors of MIPS.

This PR also adds a boolean, `meta.requiresBuildCanExecuteHost`, to signal a very common cross-compilation problem: build processes that assume the build platform can execute host platform binaries.

This PR also adds `stdenv.isAvailable pkg` which checks availability constraints for all three platforms (build, host, and target) as well as `meta.requiresBuildCanExecuteHost -> buildPlatform.canExecute hostPlatform`.  This should be preferred over `meta.availableOn platform pkg`.

#### Concrete example

One use-case for this is being able to automatically compile all eligible packages using the `mips64n32` ABI (which cuts memory usage dramatically for pointer-intensive data structures) and build on `mips64n64` only those few packages which don't support `n32` (usually because they're written in Rust or use [`boost-context`](https://github.com/NixOS/nixpkgs/pull/245300)).

The clumsy way of working around this would be to [have `buildRustPackage`    inject `meta.badPlatforms` values into the packages it creates](https://github.com/NixOS/nixpkgs/pull/245304), but    that covers only Rust (not other LLVM-based compilers) and even then covers only    those which use `buildRustPackage`.  Really, it's the wrong layer at    which to express these things.

`meta.broken` isn't helpful here either.  It is shallow, and its "deep" version `meta.available` is based on `throw`, so it can't be used to do anything other than to abort eval (see example below).  This is why `meta.availableOn` doesn't (and can't ever) understand `meta.broken`.

###### Includes

- https://github.com/NixOS/nixpkgs/pull/246644


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Commit:
stdenv: add `stdenv.isAvailable pkg`

This commit adds `stdenv.isAvailable pkg`, which checks `meta`
availability not only for the hostPlatform, but also for the
`buildPlatform` and `targetPlatform` using
`lib.meta.checkAvailability`.

Our current meta.{platforms,badPlatforms} scheme isn't capable of
expressing constraints that apply to anything other than the
hostPlatform.  For example:

- LLVM can't be used with targetPlatform.isMips64n32, but GCC can

The clumsy way of working around this is to have `buildRustPackage`
inject `meta.badPlatforms` values into the packages it creates, but
that covers only Rust (not other LLVM-based compilers) and only
those which use `buildRustPackage`.  Really, it's the wrong layer at
which to express these things.

This commit adds `meta.availability.{build,host,target}.{bad,only}`
which are computed automatically based on a package's dependencies.

This is done in a backwards-compatible way for the hostPlatform:
`meta.availability.host` will incorporate
`meta.{platforms,badPlatforms}` from both the package as well as all
of its `buildInputs`.  The constraints are cumulative, so deep
recursion (like checkMetaRecursively) is not needed; each package's
`meta.availability` only needs to look at that package's immediate
inputs.

This commit also causes check-meta.nix to utilize the information in
meta.badBuildPlatforms and meta.badTargetPlatforms.

In particular, the following test commands now behave correctly:

  # XFAIL: golang can't build *on* MIPS, but it can cross compile to it
  nix-instantiate . -A pkgsCross.mipsel-linux-gnu.go

  # SUCCESS: golang can't build *on* MIPS, but it can cross compile to it
  nix-instantiate . -A pkgsCross.mipsel-linux-gnu.dnscrypt-proxy2

  # XFAIL: golang can't target Mips64n32
  nix-instantiate . -A pkgsCross.mips64el-linux-gnuabin32.dnscrypt-proxy2

  # XFAIL: rust can't target Mips64n32
  nix-instantiate . -A pkgsCross.mips64el-linux-gnuabin32.tiny

amjoseph-nixpkgs commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -437,6 +440,50 @@ let
+      # derivation is guaranteed not to build and therefore should
+      # not even be attempted.
+      availability = let meta = attrs.meta or {}; in {
+        build = {
+          bad =
+            meta.availability.build.bad or [] ++
+            lib.concatMap (pkg: pkg.meta.availability.host.bad or []) referencesOnBuild ++
+            lib.concatMap (pkg: pkg.meta.availability.build.bad or []) referencesOnHost;
+        };
+        host = {
+          bad =
+            meta.availability.host.bad or [] ++
+            lib.concatMap (pkg: pkg.meta.availability.target.bad or []) referencesOnBuild ++
+            lib.concatMap (pkg: pkg.meta.availability.host.bad or []) referencesOnHost ++
+            meta.badPlatforms or []
+          ;
+
+          # "host.only" is needed to represent the (deprecated,
+          # legacy) meta.platforms attribute, which should be
+          # avoided in favor of meta.hydraPlatforms and
+          # meta.badPlatforms, since in order to set meta.platforms
+          # correctly you need to be able to see into the future and
+          # know what platforms nixpkgs might someday support.
+          # There are no build.only or target.only attrsets because
+          # we (thankfully) don't have meta.platforms equivalents
+          # for those.
+          only =
+            (if meta?platforms && meta.platforms != lib.platforms.all
+             then [ meta.platforms ]
+             else [ ])
+            ++ lib.concatMap (pkg: pkg.meta.availability.host.only or []) referencesOnHost;
+        };
+        target = {
+          bad =
+            meta.availability.target.bad or [] ++
+            lib.concatMap (pkg: pkg.meta.availability.target.bad or []) referencesOnHost;
+        };
+      };
```
@Artturin, I would appreciate your feedback on whether or not I got the phase offsets right here.  I think I did but I'm not super-confident of that.

amjoseph-nixpkgs commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -437,6 +445,52 @@ let
+      # derivation is guaranteed not to build and therefore should
+      # not even be attempted.
+      availability = let meta = attrs.meta or {}; in {
+        build = {
+          bad = internStringifiable (
+            lib.concatMap (pkg: pkg.meta.availability.host.bad or []) referencesOnBuild ++
+            lib.concatMap (pkg: pkg.meta.availability.build.bad or []) referencesOnHost ++
+            meta.badBuildPlatforms or []
+          );
+        };
+        host = {
+          bad = internStringifiable (
+            lib.concatMap (pkg: pkg.meta.availability.target.bad or []) referencesOnBuild ++
+            lib.concatMap (pkg: pkg.meta.availability.host.bad or []) referencesOnHost ++
+            meta.badPlatforms or []
+          );
+
+          # "host.only" is needed to represent the (deprecated,
+          # legacy) meta.platforms attribute, which should be
+          # avoided in favor of meta.hydraPlatforms and
+          # meta.badPlatforms, since in order to set meta.platforms
+          # correctly you need to be able to see into the future and
+          # know what platforms nixpkgs might someday support.
+          # There are no build.only or target.only attrsets because
+          # we (thankfully) don't have meta.platforms equivalents
+          # for those.
+          only = internStringifiable (
+            (if meta?platforms && meta.platforms != lib.platforms.all
+             then [ meta.platforms ]
+             else [ ])
+            ++ lib.concatMap (pkg: pkg.meta.availability.host.only or []) referencesOnHost
+          );
+        };
+        target = {
+          bad = internStringifiable (
+            lib.concatMap (pkg: pkg.meta.availability.target.bad or []) referencesOnHost ++
+            meta.badTargetPlatforms or []
+          );
+        };
```
@Artturin, I would appreciate your feedback on whether or not I got the phase offsets right here. I think I did but I'm not super-confident of that.

amjoseph-nixpkgs commented:
Investigating the [darwin rebuilds](https://gist.github.com/GrahamcOfBorg/53ea1a8789f5076897af88a26c392f00) this causes.  No rebuilds on `*-linux`.

piegamesde reviewed (commented):
Since it looks like you touched that code so you might know, why is `assertValidity` having both `attrs` and `meta` parameters? What's the difference between `attrs.meta` and `meta`?

amjoseph-nixpkgs commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -453,4 +507,16 @@ let
+  # deterministic order.
+  #
+  # TODO: possibly move into /lib/
+  internStringifiable = list:
```
Nix attrsets are sorted (natively, in C++ code) -- my goal here was to take advantage of that.  Maybe I misunderstood the question though?

piegamesde replied:
If this function can easily be expressed using existing `lib` functions then this should be done, and any optimizations done here should be put into lib instead. Maybe see also #239722 and #244187

amjoseph-nixpkgs replied:
Note: the elements of the list are not all strings.  There are patterns (see `lib.systems.inspect`).  Maybe that clarifies the reason for the `builtins.toJSON`.

amjoseph-nixpkgs replied:
> If this function can easily be expressed using existing `lib` functions

Using `lib.unique` to remove the duplicates from a list is O(n^2).

> Maybe see also #239722

That function is O(n log n), but it merely returns a boolean indicating if all the elements are unique, rather than removing the duplicates.  More importantly, it expects a comparison function.  In this case the needed function would be a lexicographic comparison function on attrsets, which I don't believe exists in `lib`.  Are you asking me to write one?

piegamesde replied:
Isn't `<` already implemented on most everything in Nix?

Also the discussion in there talks about adding functions which return the sorted unique list instead of a boolean, this feels like what you would want.

amjoseph-nixpkgs replied:
> Isn't `<` already implemented on most everything in Nix?

```
$ nix eval --expr '{ a=3; } < { b=2; }'
error: cannot compare a set with a set
```

piegamesde replied:
@pennae I think you told me this, did I misunderstand something?

amjoseph-nixpkgs replied:
FWIW I think a `builtins` to do lexicographic comparison on *function-free* attrsets would be a great addition to Nix.  But it isn't there yet.

amjoseph-nixpkgs replied:
@piegamesde I believe this is addressed by 1c68dedefcb5ac3c2b2757fc076a1f9850c4c1a1.  I would appreciate your review of that commit.

amjoseph-nixpkgs commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -304,6 +304,11 @@ let
     outputsToInstall = listOf str;
     position = str;
     available = unspecified;
+    availability = unspecified;
```
> Because then I don't think we really need to add this to this list

I tried removing this and the meta check failed.

Ideally I'd like to give types for the subcomponents of `availability` (see the following four lines which are commented out) -- but apparently `lib.types` doesn't have "struct types" like yants does?  Or maybe I just can't figure out how to express them.

> At the very least, this should not be located in the "Weirder stuff that doesn't appear in the documentation?" section

If this PR is merged **I will absolutely update the documentation.**

I'd like to do that in a separate PR because our documentation standards are much higher now (hooray!) and mixing together feedback on the code vs feedback on the documentation tends to make both discussions harder to follow.

piegamesde replied:
Yeah I've been struggling with this a lot too when trying to implement RFC 127. `lib.types` has clumsy struct types in form of submodules, however checking them is a pain (it `throw`s, and we have to `tryEval` to get around that, at the cost of losing any helpful error messages). See https://github.com/NixOS/nixpkgs/pull/177272#issuecomment-1621946623 for more details.

The solution I'm tentatively going for is https://github.com/NixOS/nixpkgs/pull/177272/files#diff-8d6de52a69f8de7972d373a09d7a0747f88017b71242cd7bc241d61dd431101dR468-R537, I could factor out the changes to `checkMetaAttr` if you want to make use of that as well.

amjoseph-nixpkgs replied:
> For now, maybe just move those lines into a section commented with "automatically generated in check-meta" or something

Implemented in 4a593cdd0542893b16934cc7553c1cde95b6a792

amjoseph-nixpkgs commented:
> Since it looks like you touched that code so you might know, why is `assertValidity` having both `attrs` and `meta` parameters? What's the difference between `attrs.meta` and `meta`?

Oh man was this frustrating.

`attrs` is whatever was passed to `mkDerivation`.  It contains only what was passed in to `mkDerivation`, not anything that gets added to it.

`meta` is the final `meta` of the resulting derivation; it contains everything.

So for example, `attrs.meta.validity` doesn't (or shouldn't!) exist.  `meta.validity` does exist.

The reason why `attrs` has to be passed in is to prevent infinite recursion (which I was bitten by several times while writing this PR).  While you're creating the final `meta` attrset sometimes you need to reference "the `meta` that was passed to `mkDerivation`, prior to anything being added to it" because if you use the final `meta` you'll end up depending on your own results.

This is mainly because `check-meta.nix` uses a lot of `} // if x then {` constructions, and Nix attrsets are lazy in their attrvalues *but are strict in their attrnames*.

Artturin commented on pkgs/stdenv/generic/default.nix:
```diff
@@ -164,6 +164,11 @@ let
 
       inherit cc hasCC;
 
+      isAvailable = pkg:
+        lib.meta.checkAvailability buildPlatform  pkg.meta.availability.build &&
+        lib.meta.checkAvailability hostPlatform   pkg.meta.availability.host &&
+        lib.meta.checkAvailability targetPlatform pkg.meta.availability.target;
```
I don't understand why this is desirable.

So far we have been trying to deprecate the various `stdenv.is*` in favor of the more accurate `stdenv.*Platform.is`.

amjoseph-nixpkgs replied:
> I don't understand why this is desirable.

Because availability isn't a function of *just* the `hostPlatform` (or *just* the `buildPlatform`, or *just* the `targetPlatform`).  You need all three in order to decide.

Perhaps a more-perfect solution would be to have `stdenv.platforms.{build,host,target}` and then add `stdenv.platforms.isAvailable`.  That would place `isAvailable` the smallest attrset containing all the needed information.

Artturin commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -232,6 +232,8 @@ else let
 
+  referencesOnBuild = nativeBuildInputs ++ propagatedNativeBuildInputs;
+  referencesOnHost = buildInputs ++ propagatedBuildInputs;
   references = nativeBuildInputs ++ buildInputs
```
`references` was added in https://github.com/nixos/nixpkgs/commit/eaee2a119916b2761d08cb8f68f1b4afbde15e51

> This option makes `meta.evaluate` into a close approximation of the result of
evaluating `.outPath` by checking all the dependencies recursively at a cost of
2x slowdown. Note that actually evaluating `.outPath` costs some
5x-7x more because `.outPath` also computes all the hashes.

This pr has big performance impacts too
![image](https://github.com/NixOS/nixpkgs/assets/56650223/0c2ba7ee-6a5a-4aa4-895c-0493b150c301)

amjoseph-nixpkgs replied:
> This pr has big performance impacts too (current ofborg eval time is ~11 minutes)

Percentages are more useful here: ~5% on all but one metric (and that one metric at +10% is **not** an allocation metric).

That said, there are still lots of optimization opportunities.  If performance is the only remaining obstacle I will implement those.  I believe the overhead can be brought down to ~1% but I don't want to put in that effort if it's going to be wasted.

amjoseph-nixpkgs replied:
1c68dedefcb5ac3c2b2757fc076a1f9850c4c1a1 should improve this.  Let's see what OfBorg says.

amjoseph-nixpkgs replied:
Latest eval report, at 4141be0.  It's basically neutral; a mixed bag of under-5% changes (about half faster, half slower).  The only thing that stands out is the +12% on `nrLookups`, but attrset lookups are fast and allocate no memory -- this metric is far less important than most of the others.

![eval-report](https://github.com/NixOS/nixpkgs/assets/54836058/e38a5c67-631e-445d-aa7c-b98590206a9e)

amjoseph-nixpkgs commented:
Squashed.

Commit:
rustc: badTargetPlatforms: add isMips64n32

Commit:
go: add isMips64n32 to badTargetPlatforms, isMips to badHostPlatforms

We can't build a `go` compiler on MIPS due to lack of bootstrap
binaries.  However, go is in fact capable of emitting code for MIPS,
so we can cross compile to a MIPS target (useful for embedded
routers!).

Commit:
apparmor: fix invalid reference when withPython=false

This commit fixes a bug exposed by #25322.

Commit:
xdg-utils: set meta.requiresBuildCanExecuteHost, stop abusing meta.broken

Commit:
emacs: set meta.requiresBuildCanExecuteHost, stop abusing meta.broken

Commit:
recoll: set meta.requiresBuildCanExecuteHost, stop abusing meta.broken

amjoseph-nixpkgs commented:
Once OfBorg finishes eval I'll push four more commits that use `stdenv.isAvailable` to replace abuses of `meta.broken` for `p11-kit`, `emacs`, `recoll`, and `xdg-utils`, all of which require that the build platform is able to execute `hostPlatform` binaries.

amjoseph-nixpkgs commented:
```
make[2]: *** [aws-cpp-sdk-core-tests/CMakeFiles/aws-cpp-sdk-core-tests.dir/build.make:883: aws-cpp-sdk-core-tests/aws-cpp-sdk-core-tests] Error 1
make[2]: *** Deleting file 'aws-cpp-sdk-core-tests/aws-cpp-sdk-core-tests'
make[1]: *** [CMakeFiles/Makefile2:396: aws-cpp-sdk-core-tests/CMakeFiles/aws-cpp-sdk-core-tests.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Pretty sure that's not my fault.

@ofborg eval

amjoseph-nixpkgs commented:
It turns out that most of the eval overhead is due to people making wacky choices for `meta.platforms` (sometimes `isGnu`, sometimes `isLinux`, sometimes `isFoundInMyMenagerie`, etc).  And these lists can't be unioned like `bad*Platforms` can, so they just accumulate uncontrollably.

`meta.platforms` really isn't useful for anything except as the default value of `hydraPlatforms`.  I'm going to leave `meta.platforms`' current shallow behavior and encourage people to switch to `meta.hydraPlatforms`-plus-`badPlatforms`

Commit:
stdenv: use shallow checking for meta.platforms (deep for bad*Platforms)

amjoseph-nixpkgs commented:
Latest eval report, at 4141be0.  It's basically neutral; a mixed bag of under-5% changes (about half faster, half slower).  The only thing that stands out is the +12% on `nrLookups`, but attrset lookups are fast and allocate no memory -- this metric is far less important than most of the others.

![eval-report](https://github.com/NixOS/nixpkgs/assets/54836058/e38a5c67-631e-445d-aa7c-b98590206a9e)

amjoseph-nixpkgs commented:
Next push is an experiment with having hashing replace sorting, to see which wins.

bew commented:
The PR is still in draft, what is missing before being ready for review?

bew commented:
Why did you closed it?
Is it superseded by anything else?

piegamesde commented:
https://github.com/NixOS/nixpkgs/issues/50105#issuecomment-1902682831 somebody else will have to pick up the work probably

roberth commented:
Adam has closed many of his PRs. I don't think that was necessary, and I hope he's doing well.

amjoseph-nixpkgs commented:
> I hope he's doing well.

I'm fine, Robert.  Thanks for your concern.
